### ISSUE TITLE
Any real learning happening for DCBTrainer?

### ISSUE THREAD

**Issue by TMorville** [SforAiDl/genrl#314]
I have been playing around with the DCBTrainer and found some potential inconsistencies.

1) **StatlogData** example found [here](https://genrl.readthedocs.io/en/latest/usage/tutorials/bandit/contextual_overview.html)

```
from genrl.utils import StatlogDataBandit

bandit = StatlogDataBandit(download=True)
context = bandit.reset()

from genrl.agents import NeuralLinearPosteriorAgent

agent = NeuralLinearPosteriorAgent(bandit)
context = bandit.reset()

action = agent.select_action(context)
new_context, reward = bandit.step(action)

from genrl.trainers import DCBTrainer

trainer = DCBTrainer(agent, bandit)
trainer.train(timesteps=1000, batch_size=32)
```

and code to evaluate

```
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report, accuracy_score

def _evaluate(trainer, bandit):
        
    y_true = bandit.df.iloc[:, -1].to_numpy()
    
    class_distribution = bandit.df.iloc[:, -1].value_counts()
    most_freq_class = class_distribution.idxmax()
    baseline_accuracy = accuracy_score(y_true, np.resize(most_freq_class, len(bandit.df))).round(2)
    
    tensor_matrix = torch.stack([torch.LongTensor(x).float() for x in bandit.df.iloc[:, :-1].to_numpy()])    
    
    y_pred = []
    
    for i in tensor_matrix:
        y_pred.append(trainer.agent.select_action(i).item()) 
    
    print("Baseline accuracy score: {}%".format(baseline_accuracy))
    print("After {} steps accuracy is {}%".format(agent.t, accuracy_score(y_true, y_pred).round(2)))
    print("Classification report")    
    print(classification_report(y_true, y_pred))
    
    fig = plt.figure(figsize=(10, 4))
    plt.subplot(121)
    plt.plot(bandit.cum_reward_hist)
    plt.title("Cumulative reward")
    plt.subplot(122)
    plt.plot(bandit.cum_regret_hist)
    plt.title("Cumulative regret")
    plt.tight_layout()
    
    return y_true, y_pred

yt, yp = _evaluate(trainer, bandit)

Baseline accuracy score: 0.78%
After 44501 steps accuracy is 0.78%
Classification report
              precision    recall  f1-score   support

           0       0.00      0.00      0.00         0
           1       0.78      0.99      0.88     34108
           2       0.00      0.00      0.00        37
           3       0.00      0.00      0.00       132
           4       0.14      0.00      0.00      6748
           5       0.17      0.00      0.00      2458
           6       0.00      0.00      0.00         6
           7       0.00      0.00      0.00        11

    accuracy                           0.78     43500
   macro avg       0.14      0.12      0.11     43500
weighted avg       0.65      0.78      0.69     43500
```
2) **WineDataBandit** example found [here](https://genrl.readthedocs.io/en/latest/usage/tutorials/bandit/adding_data_bandit.html)

Define the bandit
```
from typing import Tuple

import pandas as pd
import torch

from genrl.utils.data_bandits.base import DataBasedBandit
from genrl.utils.data_bandits.utils import download_data


URL = "http://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data"

class WineDataBandit(DataBasedBandit):
    def __init__(self, **kwargs):
        super(WineDataBandit, self).__init__(**kwargs)

        path = kwargs.get("path", "./data/Wine/")
        download = kwargs.get("download", None)
        force_download = kwargs.get("force_download", None)
        url = kwargs.get("url", URL)

        if download:
            path = download_data(path, url, force_download)

        self._df = pd.read_csv(path, header=None)
        self.n_actions = len(self._df[0].unique())
        self.context_dim = self._df.shape[1] - 1
        self.len = len(self._df)
        
    def reset(self) -> torch.Tensor:
        self._reset()
        self.df = self._df.sample(frac=1).reset_index(drop=True)
        return self._get_context()

    def _compute_reward(self, action: int) -> Tuple[int, int]:
        label = self._df.iloc[self.idx, 0]
        r = int(label == (action + 1))
        return r, 1

    def _get_context(self) -> torch.Tensor:
        return torch.tensor(
            self._df.iloc[self.idx, 1:].values,
            device=self.device,
            dtype=torch.float,
        )
```

training

```
bandit = WineDataBandit(path='/path/to/data')

from genrl.agents import NeuralLinearPosteriorAgent

agent = NeuralLinearPosteriorAgent(bandit)
context = bandit.reset()

action = agent.select_action(context)
new_context, reward = bandit.step(action)

from genrl.trainers import DCBTrainer

trainer = DCBTrainer(agent, bandit)
trainer.train(timesteps=5000, batch_size=32)
```

and evaluation

```
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import classification_report, accuracy_score

def _evaluate(df, trainer, bandit):
        
    y_true = bandit.df.iloc[:, 0].to_numpy()
    
    class_distribution = bandit.df.iloc[:, 0].value_counts()
    most_freq_class = class_distribution.idxmax()
    baseline_accuracy = accuracy_score(y_true, np.resize(most_freq_class, len(bandit.df))).round(2)
    
    tensor_matrix = torch.stack([torch.LongTensor(x).float() for x in bandit.df.iloc[:, 1:].to_numpy()])
    
    y_pred = []
    
    for i in tensor_matrix:
        y_pred.append(trainer.agent.select_action(i).item()) 
    
    print("Baseline accuracy score: {}%".format(baseline_accuracy))
    print("After {} steps accuracy is {}%".format(agent.t, accuracy_score(y_true, y_pred).round(2)))
    print("Classification report")    
    print(classification_report(y_true, y_pred))
    
    fig = plt.figure(figsize=(10, 4))
    plt.subplot(121)
    plt.plot(bandit.cum_reward_hist)
    plt.title("Cumulative reward")
    plt.subplot(122)
    plt.plot(bandit.cum_regret_hist)
    plt.title("Cumulative regret")
    plt.tight_layout()
    
    return y_true, y_pred

yt, yp = _evaluate(trainer, bandit)

Baseline accuracy score: 0.4%
After 5357 steps accuracy is 0.0%
Classification report
              precision    recall  f1-score   support

           0       0.00      0.00      0.00       0.0
           1       0.00      0.00      0.00      59.0
           2       0.00      0.00      0.00      71.0
           3       0.00      0.00      0.00      48.0

    accuracy                           0.00     178.0
   macro avg       0.00      0.00      0.00     178.0
weighted avg       0.00      0.00      0.00     178.0

```

For both cases (and the third titanic case referenced in #301 ) when training both reward and regret increases which could point to no actual learning happening and the increase in reward comes purely from randomly guessing instead of learning. 

Notice that for the Statlog data, the evaluation col is the **last** column, while in the wine data, its the **first** column.

**Comment by tomonodes** [comment]
tldr; parameters `mu` `a` and `b` used in the bayesian regression are not updated in training.

new class `ParameterDB` that collects parameters in training

```
class ParameterDB(object):
    """
    Database for storing parameters (mu, a, b).

    Args:
        device (str): Device to use for tensor operations.
            "cpu" for cpu or "cuda" for cuda. Defaults to "cpu".

    Attributes:
        db (dict): Dictionary containing list of parameters.
        db_size (int): Number of parameters stored in database.
        device (torch.device): Device to use for tensor operations.
    """

    def __init__(self, device: Union[str, torch.device] = "cpu"):

        if type(device) is str:
            self.device = (
                torch.device(device)
                if "cuda" in device and torch.cuda.is_available()
                else torch.device("cpu")
            )
        else:
            self.device = device

        self.db = {"mu": [], "a": [], "b": []}
        self.db_size = 0

    def add(self, mu: torch.Tensor, a: torch.Tensor, b: torch.Tensor):
        """Add (context, action, reward) transition to database

        Args:
            context (torch.Tensor): Context recieved
            action (int): Action taken
            reward (int): Reward recieved
        """
                
        self.db["mu"].append(mu)
        self.db["a"].append(a)
        self.db["b"].append(b)
        self.db_size += 1
```

New update method in `NeuralLinearPosteriorAgent`
```
    def update_param_db(self, mu: torch.Tensor, a: torch.Tensor, b: torch.Tensor):
        """
        """
        self.param_db.add(mu, a, b)
```

add update to training in `DCBTrainer`

```
        try:
            for t in range(1, timesteps + 1):
                action = self.agent.select_action(context)
                new_context, reward = self.bandit.step(action)
                self.agent.update_db(context, action, reward)
                context = new_context

                if train_epochs_schedule is not None and t < train_epochs_decay_steps:
                    train_epochs = int(train_epochs_schedule[t])

                if t > update_after and t % update_interval == 0:
                    self.agent.update_params(
                        action, kwargs.get("batch_size", 64), train_epochs
                    )
                    
                    self.agent.update_param_db(self.agent.mu, self.agent.a, self.agent.b)

                if t > ignore_init:
                    regret_mv_avgs.append(np.mean(self.bandit.regret_hist[-mv_len:]))
                    reward_mv_avgs.append(np.mean(self.bandit.reward_hist[-mv_len:]))
                    if t % log_every == 0:
                        self.logger.write(
                            {
                                "timestep": t,
                                "regret/regret": self.bandit.regret_hist[-1],
                                "reward/reward": reward,
                                "regret/cumulative_regret": self.bandit.cum_regret,
                                "reward/cumulative_reward": self.bandit.cum_reward,
                                "regret/regret_moving_avg": regret_mv_avgs[-1],
                                "reward/reward_moving_avg": reward_mv_avgs[-1],
                            }
                        )
```

running this for the Statlog example yields constant parameters through the entire training run.

See full gist [here](https://gist.github.com/TMorville/064639a0a367680e6affe9a89c782455).

**Comment by sampreet-arthi** [comment]
Is this still an issue?

**Comment by TMorville** [comment]
It is. However, I think my baseline might be wrong. 

I think the relevant baseline to compare performance to should be a bayesian regression trained directly on the data, instead of the output from the neural network. Do you agree?
